app/components/avo/index/resource_table_component.rb

```diff
@@ -13,6 +13,30 @@ def initialize(resources: nil, resource: nil, reflection: nil, parent_record: ni
     @pagy = pagy
     @query = query
     @actions = actions
+
+    cache_if Avo.configuration.cache_resources_on_index_view, resource.cache_hash(parent_record), expires_in: 1.day do
+      # Keep unique header fields, builded by joining all row visible fields.
+      @header_fields = []
+
+      # Rows keep each Avo::Index::TableRowComponent initialized to be rendered
+      # It is initialized here in order to know what fields are visible to render them as column on header.
+      @rows = @resources.map do |resource|
+        # Store each row visible fields on @header_fields variable
+        @header_fields.concat row_fields = resource.get_fields(reflection: reflection, only_root: true)
+
+        {
+          resource: resource,
+          fields: row_fields,
+          reflection: reflection,
+          parent_record: parent_record,
+          parent_resource: parent_resource,
+          actions: actions
+        }
+      end
+
+      # Render only uniq fields on header
+      @header_fields.uniq! { |field| field.table_header_label }
+    end
```

Comment:
I feel like this is a bit too much.
Can we split it or refactor it to be more easily understood at first glance?

It took a while and some concentrating power to make sure I understand what it does.

Reply:
Totally understand, let«s refactor it

Reply:
Can you please look at the last version? I tried to comment each line and split the logic